### DISCUSSION TITLE
What is the difference between listen with invalidateSelf and watch?

### DISCUSSION THREAD

**Discussion by YousefAK009** [rrousselGit/riverpod#3198] [Q&A] [answered]
I encountered a CircularDependencyError, so I got the idea to use listen with invalidateSelf instead of using the watch method. I added an invalidateSelf method inside listen to simulate the behavior of watch.

Is there any problem with this approach?

![image](https://github.com/rrousselGit/riverpod/assets/46021584/52b42b5d-88ef-4570-955e-3489461b4c0a)

**Comment by rrousselGit** [comment, marked as answer]
If you have a circular dependency error, using ref.listen instead shouldn't have any impact.
If that makes your error disappear, chances are it's a bug. I wouldn't rely on it.

But overall yes, ref.watch could be considered as ref.listen + invalidateSelf.

The only difference is how it impacts loading states ("reload" for watch vs "refresh" for invalidateSelf)

**Reply by YousefAK009** [reply]
Thanks, I'm very glad that I can use listen with invalidateSelf. It behaves as watch does, which is great because I have many use cases for it.
